### PR TITLE
ARROW-14584: [Python][CI] Python sdist installation fails with latest setuptools 58.5

### PR DESCRIPTION
**Pull request by kszucs**
This patch should go to the 6.0.1 release.

### CODE REVIEW

**Comment by kszucs** [comment]
@github-actions crossbow submit python-sdist

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by jorisvandenbossche** [comment]
Hmm, so this doesn't actually help for the sdist installation. That's because it is already failing in the step of installing numpy in the build environment (it is in this environment that we now install setuptools < 58.5).

One other difference I notice compared to a previous nightly build of python-sdist that still worked (https://github.com/ursacomputing/crossbow/runs/4089110526?check_suite_focus=true) is that it now tries to install numpy from source (   "Collecting numpy==1.16.6     Downloading numpy-1.16.6.zip (5.1 MB)"), while previously it didn't do that? And it's this building of numpy from source that now fails with the latest setuptools.

**Comment by kszucs** [comment]
Yes, just looking at why it tries to install numpy from source.

**Comment by kszucs** [comment]
We have pinned that in pyproject.toml:

```python
[build-system]
requires = [
    "cython >= 0.29",
    "numpy==1.16.6; python_version<'3.9'",
    "numpy==1.19.4; python_version=='3.9'",
    "numpy==1.21.3; python_version>'3.9'",
    "setuptools < 58.5",  # ARROW-14584
    "setuptools_scm",
    "wheel"
]
```

**Comment by kszucs** [comment]
> But the fact that the requirement is "numpy>=1.16.6" seems a bit fishy, since the pyproject.toml clearly indicates "==1.16.6".

Could be due to different setuptools versions. 

I wonder what's the exact purpose of `build-system` definition since numpy `1.16.6` doesn't have wheels for `python > 3.7` which makes it pretty inconvenient to compile pyarrow from `sdist` for those python versions. 

Perhaps we should change the numpy constraints to use ">=" instead of "=".

**Comment by kszucs** [comment]
@github-actions crossbow submit python-sdist

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by jorisvandenbossche** [comment]
> Perhaps we should change the numpy constraints to use ">=" instead of "=". Going to try it.

No, we shouldn't do that. Numpy is only forward compatible on a binary level, so you need to compile against the oldest version (in the build environment it is not known what will be the numpy version at runtime).

> I wonder what's the exact purpose of `build-system` definition since numpy `1.16.6` doesn't have wheels for `python > 3.7` which makes it pretty inconvenient to compile pyarrow from `sdist` for those python versions.

That's because our numpy / python versions are not detailed enough. We should include another entry for Python 3.8 requiring numpy 1.17.3 (which does have wheels for Python 3.8). That might actually fix the issue here.

(and that's another reason for using oldest-supported-numpy, https://github.com/scipy/oldest-supported-numpy)

File: python/pyproject.toml [hunk 1]
```diff
@@ -18,10 +18,10 @@
 [build-system]
 requires = [
     "cython >= 0.29",
-    "numpy==1.16.6; python_version<'3.9'",
-    "numpy==1.19.4; python_version=='3.9'",
-    "numpy==1.21.3; python_version>'3.9'",
-    "setuptools",
+    "numpy>=1.16.6; python_version<'3.9'",
```
Comment (jorisvandenbossche) [inline python/pyproject.toml:21]:
```suggestion
    "numpy==1.16.6; python_version<'3.8'",
    "numpy==1.17.3; python_version=='3.8'",
```

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel python-sdist

**Comment by kszucs** [comment]
> that's another reason for using oldest-supported-numpy, https://github.com/scipy/oldest-supported-numpy)

Let's try that out https://github.com/apache/arrow/pull/11606

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-windows-cp39-amd64

**Review by pitrou — APPROVED** [review]
Looks ok on the principle.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-windows-cp310-amd64

**Pull request closed without merging**